pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go

```diff
@@ -241,13 +224,66 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	return ctrl.Result{}, r.UpdateStatus(ctx, app)
 }
 
-// if any finalizers newly registered, return true
-func registerFinalizers(app *v1beta1.Application) bool {
-	if !meta.FinalizerExists(&app.ObjectMeta, resourceTrackerFinalizer) && app.Status.ResourceTracker != nil {
-		meta.AddFinalizer(&app.ObjectMeta, resourceTrackerFinalizer)
-		return true
+func (r *Reconciler) handleFinalizers(ctx context.Context, app *v1beta1.Application) (bool, error) {
+	if app.ObjectMeta.DeletionTimestamp.IsZero() {
+		// NOTE Because resource tracker is cluster-scoped resources, we cannot garbage collect them
+		// by setting application(namespace-scoped) as their owner. So we must delete all
+		// resource trackers through app controller's finalizer logic.
+
+		// 'resourceTrackerFinalizer' is used to delete the resource tracker of the last app revision
+		if !meta.FinalizerExists(&app.ObjectMeta, resourceTrackerFinalizer) {
+			meta.AddFinalizer(&app.ObjectMeta, resourceTrackerFinalizer)
+			klog.InfoS("Register new finalizer for application", "application", klog.KObj(app), "finalizer", resourceTrackerFinalizer)
+			return true, errors.Wrap(r.Client.Update(ctx, app), errUpdateApplicationFinalizer)
+		}
+		// 'onlyRevisionFinalizer' is used to delete all resource trackers of app revisions which
+		// may be used out of the domain of app controller, e.g., AppRollout controller.
+		if app.Annotations[oam.AnnotationAppRevisionOnly] == "true" ||
+			len(app.Annotations[oam.AnnotationAppRollout]) != 0 || app.Spec.RolloutPlan != nil {
+			if !meta.FinalizerExists(&app.ObjectMeta, onlyRevisionFinalizer) {
+				meta.AddFinalizer(&app.ObjectMeta, onlyRevisionFinalizer)
```

Comment:
why do we need a finalizer for revision mode?

Reply:
If someone use revision to dispatch resources, and we delete the application, we need this finalizer to GC the resource trackers created according an app revision.

Reply:
I remove finalizer for revision mode and keep it for rollout mode.  I'm not sure whether appDeployment will emit resources in control-plane cluster. If so, we need this finalier to garbage collect resources while deleting an application.